### PR TITLE
signal-desktop: Add xdg-utils/bin to PATH

### PR DESCRIPTION
**Pull request by x10an14**
Since:
1. clicking on links doesn't open tab/window in default browser (tested
   on Sway),
2. it's got a hard-coded $PATH reference to `xdg-open` somewhere in
   binary (see below).

When `signal-desktop` is invoked in a terminal, the output below
appears each time a link is clicked on in-app:
```
(...)
LaunchProcess: failed to execvp:
xdg-open
```

###### Motivation for this change
See above.
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by x10an14** [comment]
Tagging listed maintainers to see if they'd like to give it a thumbs up/down/feedback: @ixmatus @primeos @equirosa 

Things to keep in mind:
- https://discourse.nixos.org/t/making-xdg-open-more-resilient/16777/5
- Slack's package does the same as done in this PR
- Search results of "discord link" in nixpkgs issues seem to always conclude with "install xdg-utils".

Ref. last point; I wrote this PR in the spirit of adhering to the hermetic philosophy of Nix/NixOS.

**Comment by iblech** [comment]
Very cool! Thanks to tackling this! I cannot test this right now but the philosophy sure is solid and the commit looks good to me.

**Review by primeos — CHANGES_REQUESTED** [review]
Not sure about this. IIRC we once decided against this (it's a known issue at least) but we're already doing it for some other packages like Chromium so it should be acceptable. I'd like to get a second opinion on it though.
**Edit:** Nvm, seems like I'm guilty of adding it for Chromium (https://github.com/NixOS/nixpkgs/pull/96922 but IIRC it was due to requests from users) and given that I didn't see any objections back then it should definitely be fine for Signal-Desktop as well.

The reasoning against this change is that a lot of tools/applications depend on xdg-open and it should be installed user/systems-wide, if desired (otherwise we'll have to patch most Electron apps, etc.).

**Comment by x10an14** [comment]
@primeos
>  (...) The reasoning against this change is that a lot of tools/applications depend on xdg-open and it should be installed user/systems-wide, if desired (otherwise we'll have to patch most Electron apps, etc.).

Thanks for enlightening me/us =)
Just to make sure I understood correctly, it's a reasoning based on the "ugh, that'll be so much work" rationale, rather than something else?

I'm asking to ensure I understood the reasoning correctly, instead of misinterpreting^^

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix
Comment (x10an14) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix]: Done! Force-pushed changes.

**Comment by jtojnar** [comment]
One issue I see is that `xdg-open` may not be sufficient – based on what your system, you might also need stuff like `gio` or `kio` programs, `shared-mime-info`, etc. Modern apps have many expectations what is available (especially around various XDG services). Desktop environments like GNOME Shell or Plasma should have you covered but when using Sway, you are on your own. That is why I would rather add NixOS documentation on how to configure your system to be “XDG compliant” (even though there is not any concrete thing). We could even add a XDG base module but then I am not sure people would want to enable it.

**Comment by x10an14** [comment]
> @jtojnar One issue I see is that `xdg-open` may not be sufficient (...)
> That is why I would rather add NixOS documentation on how to configure your system to be “XDG compliant” (even though there is not any concrete thing). We could even add a XDG base module but then I am not sure people would want to enable it.

That's a fair and valid point you raise. For the sake of my own understanding though (and honestly not for the sake of arguing for this specific PR), how do you square that with the `hermetically sealed apps` part of Nix's philosophy?

**Comment by jtojnar** [comment]
> That's a fair and valid point you raise. For the sake of my own understanding though (and honestly not for the sake of arguing for this specific PR), how do you square that with the `hermetically sealed apps` part of Nix's philosophy?

I am afraid that is not achievable for modern desktop apps without some kind of sandboxing. Almost everything depends on some D-Bus or systemd service so we would need something like what Flatpak does with `xdg-dbus-proxy`.

So I would say that part of philosophy is more “marketing speak” than anything that works in today’s reality. But another important pillar of Nix is pragmatism, and viewed through that optics, it can be said to be “hermetic enough” – at least for domains other than desktop environments.

**Comment by x10an14** [comment]
@jtojnar Thanks for answering! =) Then I guess the fate of this PR will decide whether or not I install `xdg-utils` or not x)

If this PR is accepted, I intend to propose another identical PR for Discord.

**Comment by primeos** [comment]
> Just to make sure I understood correctly, it's a reasoning based on the "ugh, that'll be so much work" rationale, rather than something else?

No, sorry, I was lazy and explained it too briefly. We really should open a dedicated issue for this (if there isn't already one but from a quick search before writing my previous comment I didn't find one). If someone has time and motivation that'd be great.

I meant that it's more of a system dependency (something that the desktop environment provides). @jtojnar explained it really well (and better than I could have as I lack some knowledge in that area - I even considered pinging you but didn't dare to bother you with it unless we have a dedicated issue for it), thanks a lot for that! :)
I fully agree that this should be documented and ideally even covered via a NixOS module. That be much better than the current approach of patching popular packages (apart from requiring a lot of effort such fixes will often be incomplete and the implementations will often diverge - and at least to me it doesn't feel like the right place to address this issue).

> how do you square that with the hermetically sealed apps part of Nix's philosophy?

There are unfortunately limits. Another interesting view to consider is that xdg-utils alone isn't enough. A web browser will also be required (has to be in `$PATH` too). But we don't want to bundle a browser like Firefox/Chromium as well, right? (And of course the web browser might also require D-Bus, etc.).

> And to be clear, if the maintainer of the package feels this patch is important, I have no issue with it. Again, pragmatism is a core value of Nix.

I'm not keen on this patch tbh. I'm not rejecting it / objecting but I'd much prefer a more general solution (documentation + NixOS module).

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix
Comment (ehmry) [inline pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:127]: I should be `--suffix` not `--prefix`. I and at least a few others write our own `xdg-open` scripts. When this is `--prefix` we would need to inject our custom `xdg-open` using an overlay and then rebuild all package like this, whereas with `--suffix` we can override at runtime.

Comment (x10an14) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix]: Sonds good! Didn't think of that, I'll take your word for it.

Just confirmed that it it works all the same for my use-case, so commit pushed!

**Comment by x10an14** [comment]
After some more pondering on where to draw the line with "adding built-in support for `xdg-open`" vs "expecting users to install it system-wide instead", I have formed an opinion I'd like to share and receive some feedback on.

1. If a program's binary/source hard-codes references to `xdg-open` (expecting it to be in PATH), I vote for "packing it into" the nixpkgs derivation(s) in question. Ref. making it "hermetically sealable".

   (As also done in Slack here in nixpkgs).
1. I can have misunderstood/been mistaken in my previous assumption (the one immediately above), and the hard-coding of `xdg-open` is more akin to a "dbus-like invocation" (not a `$PATH` expectation). If so, then I agree that it's better to let the users decide for themselves if they want to make use of `xdg-utils`' `xdg-open` or something else.

But my 2nd. point above (which presumes that theassumptions 1. are wrong) assumes that this "dbus-like invocation is also standardized enough (the way to invoke it through the invocation of `xdg-open` that is).
Because if it isn't, I don't see any cleaner solution than just packing `xdg-open` into the derivation (for now, until that gets standardized) .

`xdg-open` was supposed to allow people to re-direct invocations to programs of their own choice, no?

Doesn't make that much sense for me for there to be a `dbus-like` system in place, unless it's specifically intended for segmented/portaled/flatpak like systems.
And if so, I argue (in my limited understanding/knowledge) that it makes more sense for flatpak and their like to "own" the responsibility of the required compatibility layer.

Anyone agree/disagree/have thoughts to share on that?

**Comment by ehmry** [comment]
I think if an application expects xdg-open to be there and to execute successfully then we should make that happen. But we should also do our best to make the implementation of xdg-open overridable in the runtime environment because the application probably doesn't depend on any specific behavior from xdg-open, which would otherwise make it a leaky abstraction.

In a better world we could have a stub xdg-open that only warns the user that they don't have an xdg-open installed in $PATH, but I don't think we could do that without creating something more complicated than xdg-open already is.

**Review by ckiee — CHANGES_REQUESTED** [review]
You seem to have this spread over two commits, squash and use the commit message that follows [`CONTRIBUTING.md`](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md)?

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix
Comment (x10an14) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:7]: Good to know, thanks! =)

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix
Comment (x10an14) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix]: Fixed, thanks! =)

**Comment by x10an14** [comment]
> You seem to have this spread over two commits, squash and use the commit message that follows [`CONTRIBUTING.md`](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md)?

Thanks! Yeah, I had plans to rebase it later depending on if this PR moves forward with the discussion here in thread^^

But I've fixed up the things you pointed out how, thanks again for the look-over! =)

**Comment by ckiee** [comment]
> thanks again for the look-over! =)

No problem (:

> depending on if this PR moves forward with the discussion here in thread^^

I think it's a strawman considering you aren't setting a new precedent. If someone has a better solution than the existing one they should make a PR for that. 

```
ckie@cookiemonster ~/git/nixpkgs -> rg makeBinPath.+xdg-utils | wc -l
17
```

**Pull request merged**
